File: src/main/java/com/example/distributedprinting/client/ClientNode.java
package com.example.distributedprinting.client;

import com.example.distributedprinting.proto.*;
import com.google.common.util.concurrent.ListenableFuture;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;

import java.io.IOException;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;

public class ClientNode {
    private final int id;
    private final int port;
    private final String printerTarget;
    private final List<String> peerTargets; // e.g., host:port

    private final LamportClock clock = new LamportClock();
    private volatile State state = State.RELEASED;
    private long currentRequestTs = -1;
    private int currentRequestNum = 0;

    private final Object lock = new Object();
    private final List<Deferred> deferredQueue = new ArrayList<>();

    private Server server;
    private ManagedChannel printerChannel;
    private PrintingServiceGrpc.PrintingServiceBlockingStub printerStub;

    private final Map<String, ManagedChannel> peerChannels = new ConcurrentHashMap<>();
    private final Map<String, MutualExclusionServiceGrpc.MutualExclusionServiceFutureStub> peerFutureStubs = new ConcurrentHashMap<>();
    private final Map<String, MutualExclusionServiceGrpc.MutualExclusionServiceBlockingStub> peerBlockingStubs = new ConcurrentHashMap<>();

    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(2);

    public ClientNode(int id, int port, String printerTarget, List<String> peerTargets) {
        this.id = id;
        this.port = port;
        this.printerTarget = printerTarget;
        this.peerTargets = new ArrayList<>(peerTargets);
    }

    public void start() throws IOException, InterruptedException {
        // Start gRPC server for MutualExclusionService
        this.server = ServerBuilder.forPort(port)
                .addService(new MutualExclusionServiceImpl())
                .build()
                .start();
        System.out.println("[Client " + id + "] MutualExclusionService on port " + port);

        // Create printer client stub
        this.printerChannel = ManagedChannelBuilder.forTarget(printerTarget).usePlaintext().build();
        this.printerStub = PrintingServiceGrpc.newBlockingStub(printerChannel);

        // Create peer client stubs
        for (String target : peerTargets) {
            ManagedChannel ch = ManagedChannelBuilder.forTarget(target).usePlaintext().build();
            peerChannels.put(target, ch);
            peerFutureStubs.put(target, MutualExclusionServiceGrpc.newFutureStub(ch));
            peerBlockingStubs.put(target, MutualExclusionServiceGrpc.newBlockingStub(ch));
        }

        // Schedule automatic print jobs
        scheduler.scheduleWithFixedDelay(this::attemptPrintJob, 1000, 3000, TimeUnit.MILLISECONDS);

        // Keep running
        server.awaitTermination();
    }

    private void attemptPrintJob() {
        try {
            String message = "Mensagem do cliente " + id + " @ " + System.currentTimeMillis();
            enterCriticalSectionAndPrint(message);
        } catch (Exception e) {
            System.err.println("[Client " + id + "] Error in print job: " + e);
        }
    }

    private boolean hasPriorityOver(AccessRequest incoming) {
        // this returns true if THIS node's request has priority over incoming
        // Priority by (timestamp, id) lexicographic order: smaller wins
        if (currentRequestTs < incoming.getLamportTimestamp()) return true;
        if (currentRequestTs > incoming.getLamportTimestamp()) return false;
        return this.id < incoming.getClientId();
    }

    private void enterCriticalSectionAndPrint(String message) throws Exception {
        // Request phase
        List<ListenableFuture<AccessResponse>> futures = new ArrayList<>();
        int reqNum;
        long ts;
        synchronized (lock) {
            state = State.WANTED;
            reqNum = ++currentRequestNum;
            ts = clock.onSend();
            currentRequestTs = ts;
            System.out.printf("[Client %d] Requesting CS ts=%d req#%d%n", id, ts, reqNum);
        }

        AccessRequest req = AccessRequest.newBuilder()
                .setClientId(id)
                .setLamportTimestamp(ts)
                .setRequestNumber(reqNum)
                .build();

        for (Map.Entry<String, MutualExclusionServiceGrpc.MutualExclusionServiceFutureStub> e : peerFutureStubs.entrySet()) {
            ListenableFuture<AccessResponse> fut = e.getValue().requestAccess(req);
            futures.add(fut);
        }

        // Wait for all peer grants
        for (ListenableFuture<AccessResponse> f : futures) {
            AccessResponse ar = f.get(); // blocks until peer grants
            clock.onReceive(ar.getLamportTimestamp());
        }

        // Enter CS
        synchronized (lock) {
            state = State.HELD;
            System.out.printf("[Client %d] Entering CS ts=%d%n", id, clock.get());
        }

        // Send to printer (dumb server)
        PrintRequest printReq = PrintRequest.newBuilder()
                .setClientId(id)
                .setLamportTimestamp(clock.onSend())
                .setRequestNumber(reqNum)
                .setMessageContent(message)
                .build();
        PrintResponse printResp = printerStub.sendToPrinter(printReq);
        clock.onReceive(printResp.getLamportTimestamp());
        System.out.printf("[Client %d] Printed OK: %s%n", id, printResp.getConfirmationMessage());

        // Exit CS: grant all deferred, broadcast release
        List<Deferred> toGrant;
        synchronized (lock) {
            state = State.RELEASED;
            toGrant = new ArrayList<>(deferredQueue);
            deferredQueue.clear();
        }
        for (Deferred d : toGrant) {
            sendGrantResponse(d);
        }

        AccessRelease rel = AccessRelease.newBuilder()
                .setClientId(id)
                .setLamportTimestamp(clock.onSend())
                .setRequestNumber(reqNum)
                .build();

        // Fire-and-forget to peers
        for (MutualExclusionServiceGrpc.MutualExclusionServiceBlockingStub stub : peerBlockingStubs.values()) {
            try {
                stub.releaseAccess(rel);
            } catch (Exception ex) {
                // ignore errors on broadcast
            }
        }

        System.out.printf("[Client %d] Released CS%n", id);
    }

    private void sendGrantResponse(Deferred d) {
        try {
            long ts = clock.onSend();
            AccessResponse resp = AccessResponse.newBuilder()
                    .setAccessGranted(true)
                    .setLamportTimestamp(ts)
                    .build();
            d.observer.onNext(resp);
            d.observer.onCompleted();
        } catch (Exception e) {
            // observer may be cancelled if requester dropped
        }
    }

    private class MutualExclusionServiceImpl extends MutualExclusionServiceGrpc.MutualExclusionServiceImplBase {
        @Override
        public void requestAccess(AccessRequest request, StreamObserver<AccessResponse> responseObserver) {
            boolean defer;
            synchronized (lock) {
                clock.onReceive(request.getLamportTimestamp());
                defer = (state == State.HELD) || (state == State.WANTED && hasPriorityOver(request));
                if (defer) {
                    deferredQueue.add(new Deferred(request, responseObserver));
                    System.out.printf("[Client %d] Deferring to %d (ts=%d) state=%s%n", id, request.getClientId(), request.getLamportTimestamp(), state);
                } else {
                    long ts = clock.onSend();
                    AccessResponse resp = AccessResponse.newBuilder()
                            .setAccessGranted(true)
                            .setLamportTimestamp(ts)
                            .build();
                    responseObserver.onNext(resp);
                    responseObserver.onCompleted();
                    System.out.printf("[Client %d] Granted to %d immediately%n", id, request.getClientId());
                }
            }
        }

        @Override
        public void releaseAccess(AccessRelease request, StreamObserver<com.google.protobuf.Empty> responseObserver) {
            clock.onReceive(request.getLamportTimestamp());
            // No special action needed here per RA; releasing peer just for logging.
            System.out.printf("[Client %d] Peer %d released (req#%d) ts=%d%n", id, request.getClientId(), request.getRequestNumber(), request.getLamportTimestamp());
            responseObserver.onNext(com.google.protobuf.Empty.getDefaultInstance());
            responseObserver.onCompleted();
        }
    }

    private static class Deferred {
        final AccessRequest request;
        final StreamObserver<AccessResponse> observer;

        Deferred(AccessRequest request, StreamObserver<AccessResponse> observer) {
            this.request = request;
            this.observer = observer;
        }
    }

    public static void main(String[] args) throws Exception {
        int id = -1;
        int port = 0;
        String server = null;
        List<String> clients = new ArrayList<>();

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--id":
                    id = Integer.parseInt(nextArg(args, ++i, "--id requires value"));
                    break;
                case "--server":
                    server = nextArg(args, ++i, "--server requires value (host:port)");
                    break;
                case "--port":
                    port = Integer.parseInt(nextArg(args, ++i, "--port requires value"));
                    break;
                case "--clients":
                    String list = nextArg(args, ++i, "--clients requires value");
                    if (!list.isBlank()) {
                        clients.addAll(Arrays.asList(list.split(",")));
                    }
                    break;
            }
        }

        if (id < 0 || port == 0 || server == null) {
            System.err.println("Usage: java ClientNode --id <int> --server <host:port> --port <int> --clients <host:port,host:port,...>");
            System.exit(1);
        }

        // Remove self from peers if included
        clients.removeIf(t -> t.endsWith(":" + port));

        ClientNode node = new ClientNode(id, port, server, clients);
        node.start();
    }

    private static String nextArg(String[] args, int idx, String err) {
        if (idx >= args.length) throw new IllegalArgumentException(err);
        return args[idx];
    }
}

File: python/client_node.py
import argparse
import os
import random
import sys
import threading
import time
from concurrent import futures
from dataclasses import dataclass
from typing import Dict, List, Tuple

import grpc
from google.protobuf import empty_pb2

import printing_pb2
import printing_pb2_grpc
from lamport import LamportClock

sys.path.append(os.path.dirname(__file__))

State = type("State", (), {"RELEASED": "RELEASED", "WANTED": "WANTED", "HELD": "HELD"})

@dataclass
class Deferred:
    client_id: int
    ts: int
    req_num: int
    event: threading.Event

class MutualExclusionServicer(printing_pb2_grpc.MutualExclusionServiceServicer):
    def __init__(self, node: "ClientNode") -> None:
        self.node = node

    def RequestAccess(self, request, context):

        self.node.clock.on_receive(request.lamport_timestamp)

        with self.node.lock:
            defer = self.node.state == State.HELD or (
                self.node.state == State.WANTED
                and self.node.has_priority_over(
                    request.client_id, request.lamport_timestamp
                )
            )
            if defer:
                ev = threading.Event()
                self.node.deferred.append(
                    Deferred(
                        request.client_id,
                        request.lamport_timestamp,
                        request.request_number,
                        ev,
                    )
                )
                print(
                    f"[Client {self.node.id}] Deferring to {request.client_id} (ts={request.lamport_timestamp}) state={self.node.state}"
                )

        if defer:
            ev.wait()

        ts = self.node.clock.on_send()
        return printing_pb2.AccessResponse(access_granted=True, lamport_timestamp=ts)

    def ReleaseAccess(self, request, context):
        self.node.clock.on_receive(request.lamport_timestamp)
        print(f"[Client {self.node.id}] Peer {request.client_id} released (req#{request.request_number}) ts={request.lamport_timestamp}")
        return empty_pb2.Empty()


class ClientNode:
    def __init__(self, id_: int, port: int, printer_target: str, peer_targets: List[str]
    ) -> None:
        self.id = id_
        self.port = port
        self.printer_target = printer_target
        self.peer_targets = [t for t in peer_targets if not t.endswith(f":{port}")]

        self.clock = LamportClock()
        self.state = State.RELEASED
        self.current_req_ts = -1
        self.current_req_num = 0

        self.lock = threading.Lock()
        self.deferred: List[Deferred] = []

        self._server = None
        self._printer_channel = None
        self._printer_stub = None
        self._peer_channels: Dict[str, grpc.Channel] = {}
        self._peer_stubs: Dict[str, printing_pb2_grpc.MutualExclusionServiceStub] = {}
        self._rng = random.Random()

    def start(self):

        self._server = grpc.server(futures.ThreadPoolExecutor(max_workers=16))
        printing_pb2_grpc.add_MutualExclusionServiceServicer_to_server(
            MutualExclusionServicer(self), self._server
        )
        self._server.add_insecure_port(f"[::]:{self.port}")
        self._server.start()
        print(f"[Client {self.id}] MutualExclusionService on port {self.port}")

        self._printer_channel = grpc.insecure_channel(self.printer_target)
        self._printer_stub = printing_pb2_grpc.PrintingServiceStub(
            self._printer_channel
        )

        for target in self.peer_targets:
            ch = grpc.insecure_channel(target)
            self._peer_channels[target] = ch
            self._peer_stubs[target] = printing_pb2_grpc.MutualExclusionServiceStub(ch)

        t = threading.Thread(target=self._auto_print_loop, daemon=True)
        t.start()

        self._server.wait_for_termination()

    def _auto_print_loop(self):
        while True:
            try:
                delay = 2 + self._rng.randrange(0, 3)
                time.sleep(delay)
                msg = f"Mensagem do cliente {self.id} @ {int(time.time())}"
                self.enter_cs_and_print(msg)
            except Exception as e:
                print(f"[Client {self.id}] Error in print loop: {e}")

    def has_priority_over(self, other_id: int, other_ts: int) -> bool:

        if self.current_req_ts < other_ts:
            return True
        if self.current_req_ts > other_ts:
            return False
        return self.id < other_id

    def enter_cs_and_print(self, message: str):
        with self.lock:
            self.state = State.WANTED
            self.current_req_num += 1
            self.current_req_ts = self.clock.on_send()
            reqnum = self.current_req_num
            ts = self.current_req_ts
            print(f"[Client {self.id}] Requesting CS ts={ts} req#{reqnum}")

        req = printing_pb2.AccessRequest(
            client_id=self.id, lamport_timestamp=ts, request_number=reqnum
        )

        for target, stub in self._peer_stubs.items():
            resp = stub.RequestAccess(req)
            self.clock.on_receive(resp.lamport_timestamp)

        with self.lock:
            self.state = State.HELD
            print(f"[Client {self.id}] Entering CS ts={self.clock.value}")

        presp = self._printer_stub.SendToPrinter(
            printing_pb2.PrintRequest(
                client_id=self.id,
                message_content=message,
                lamport_timestamp=self.clock.on_send(),
                request_number=reqnum,
            )
        )
        self.clock.on_receive(presp.lamport_timestamp)
        print(f"[Client {self.id}] Printed OK: {presp.confirmation_message}")

        with self.lock:
            self.state = State.RELEASED
            to_release = list(self.deferred)
            self.deferred.clear()
        for d in to_release:
            d.event.set()

        rel = printing_pb2.AccessRelease(
            client_id=self.id,
            lamport_timestamp=self.clock.on_send(),
            request_number=reqnum,
        )
        for stub in self._peer_stubs.values():
            try:
                stub.ReleaseAccess(rel)
            except Exception:
                pass
        print(f"[Client {self.id}] Released CS")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--id", type=int, required=True)
    parser.add_argument("--server", type=str, required=True, help="printer host:port")
    parser.add_argument("--port", type=int, required=True)
    parser.add_argument("--clients", type=str, default="")
    args = parser.parse_args()

    peers = [c for c in args.clients.split(",") if c]
    node = ClientNode(args.id, args.port, args.server, peers)
    node.start()


if __name__ == "__main__":
    main()
